read support for some Japanese ASCII data formats

yannikbehr opened the pull request:
read support for the Japanese K-NET/KiK-net strong-motion seismograph waveforms
read support for the NIED(Fnet) moment tensor catalog

QuLogic commented on obspy/core/util/base.py:
```diff
@@ -37,7 +37,8 @@
                    'io.sh', 'io.shapefile', 'io.stationxml', 'io.wav',
-                   'io.xseed', 'io.y', 'io.zmap', 'realtime', 'signal', 'taup']
+                   'io.xseed', 'io.y', 'io.zmap', 'realtime', 'signal', 'taup',
+                   'io.nied', 'io.knet']
```
Please keep this list sorted.

krischer commented on obspy/io/knet/__init__.py:
```diff
@@ -0,0 +1,115 @@
+            u'last correction': UTCDateTime(1996, 8, 10, 18, 0),
+            u'evla': 38.92, u'mag': 5.9, u'accmax': 4.383, u'duration': 59.0,
+            u'evdp': 7.0, u'evot': UTCDateTime(1996, 8, 10, 18, 12),
+            u'evlo': 140.63})
```
This doctest will randomly fail on Python 3 as dictionary order is actually random on Python 3 and changes between runs. I am not aware of a nice way to test dictionaries in doctests.

yannikbehr replied:
According to this post pprint may work: http://stackoverflow.com/questions/15549429/how-do-i-test-dictionary-equality-with-pythons-doctest-package

Would that be an acceptable solution?

krischer replied:
Yea that does work but it kind of defeats the purpose of doctests a bit...suddenly it is much less readable and it could just end up in a normal test. But that is just my opinion.

Did you see my PR to your branch? I kind of worked around it a bit in there.

yannikbehr replied:
Hadn't seen this. Thanks.

krischer commented on obspy/io/nied/util.py:
```diff
@@ -0,0 +1,39 @@
+    dx = (((370 * 24) * 60) * 60) * 1000
+    rng = 1
+    tmp = rng
+    for i in xrange(numenc):
```
`xrange` does not exist on Python 3. The `from future.builtins import *` line actually make the normal `range()` behave like `xrange()` on Python 2 so just using `range` is always ok.

megies commented:
We have some pure ASCII formats in `obspy.io.ascii`.. but maybe keeping it separate is better for visibility..

yannikbehr commented:
@megies @QuLogic  @krischer Since there are a bunch of Japan specific data formats (both ascii and binary), how about bundling them in one module, e.g. obspy.io.japan?

krischer commented:
Might be a pretty good idea so +1 from my side.

megies commented:
Although `io.japan` feels a bit strange.. but bundling is good I think. So +1 I guess.

yannikbehr commented:
How about io.nied (for the institute which operates most of the networks)?

ThomasLecocq commented:
here http://www.kyoshin.bosai.go.jp/kyoshin/man/knetform_en.html

they call it the k-net data format ?

megies commented:
+1 for `io.nied` (looks like that's were all these formats come from?)

yannikbehr commented:
The two formats I included come from NIED but, for example, waveform data from Hi-net (the high-gain broad-band sensor network) is provided in the 'win32' format which was developed by another institute. Then there is also JMA (Japan Meteorological Agency) which I believe has its own catalog format. But for now, let's stick to `io.nied` and if someone adds parsers for other Japanese data formats we can revisit the naming question.

claudiodsf commented:
That's great! I had similar code but never found the time to contribute it... Glad you could do it properly!

Any plans for the binary 'win32' format (developed by ERI)? Ideally a win32 plugin could be developed starting from relevant parts of 'WIN System' (http://eoc.eri.u-tokyo.ac.jp/WIN/Eindex.html), which is under GPL.

megies commented:
> Any plans for the binary 'win32' format (developed by ERI)?

I guess there's not a chance this `win32` is connected in any way to what is currently `io.datamark`? Just asking, because @ThomasLecocq wanted to rename it to `io.win`..

claudiodsf commented:
@megies, it may be that the format used by the Datamark logger (Japanese equipment) is some sort of variation of the win format. There are quite a few Datamark-related comments in the 'Win System' source code...
Let's see if @ThomasLecocq has comments on that.

yannikbehr commented:
I just re-organized the code to bundle the two Japanese formats into an `io.nied` module. What doesn't seem to work yet is to run the tests with obspy-runtests. @krischer would you mind having a quick look at that? @megies is this the kind of structure you had in mind?

megies commented:
The docs `obspy.io.knet.rst` can be removed again, right? Not sure why `obspy-runtests` doesn't work, but I would not nest this down into two levels, I'd rather make it.. but just my personal opinion..

`io/nied/knet.py`
`io/nied/fnetmt.py`
`io/nied/tests/test_knet.py`
`io/nied/tests/test_fnetmt.py`

yannikbehr commented:
@megies I joined the tests 

Regarding running the tests, the following works:
`obspy-runtests obspy.io.nied.tests.test_fnetmt_reading`

but this doesn't
`obspy-runtests obspy.io.nied`

krischer commented:
@yannikbehr  `obspy-runtests io.nied` works fine for me...i agree that's a bit inconsistent.

yannikbehr commented:
@krischer OK, great. I was just using obspy-runtests incorrectly then.

krischer commented:
I rebased on top of the current master and force pushed. Good to merge from my point of view once and if CI passes.

megies commented:
Still think the last directory nesting brings no advantages and "Flat is better than nested."
But then again, who cares I guess. :+1:

QuLogic commented on misc/docs/source/conf.py:
```diff
@@ -13,9 +13,13 @@
 
 import obspy
 
-
```
Why?

yannikbehr replied:
Comes from 0b56c18

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+    return res_id
+
+
+def _buffer_proxy(filename_or_buf, function, reset_fp=True,
```
We really should move this into some utility section.

krischer replied:
I agree. I chose to not do it here as we much rather should get this across all modules which is already in some other ticket. Let's hope one of us finds the time to do that soon.

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+                        return False
+                    return True
+            cnt += 1
+    except:
```
What kind of exceptions are you expecting here? The only one I can see is the `int` conversion.

yannikbehr replied:
I don't think the `int` conversion can fail as the regular expression only matches (\d+). I just put the `try` `except` around it to account for anything I may have forgotten. If this is considered bad practice I'll remove it again.

QuLogic replied:
I don't really like using bare `except` because it hides bugs. So it really does depend on what's expected here.

megies replied:
:+1: for trying to avoid the general try/except, especially because the block is so long..

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+
+def _is_fnetmt_catalog(filename_or_buf):
+    """
+    Checks if the file is a F-net moment tensor file.
```
s/a/an/

yannikbehr replied:
?

QuLogic commented on obspy/io/nied/tests/test_fnetmt_reading.py:
```diff
@@ -0,0 +1,80 @@
+    def test_is_fnetmt_catalog(self):
+        """
+        This tests the _is_fnetmt_catalog method by validating that each file
+        in the data directory is a F-net catalog file and each file in the
```
s/ a / an /

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+    # First read the headerlines containing the data request parameters
+    headerlines = []
+    while True:
+        if buf.tell() >= size:
```
You could just make this test the condition of the loop, I think.

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+        # read the next event.
+        if line:
+            buf.seek(cur_pos, 0)
+            events.append(__read_single_fnetmt_entry(buf))
```
Maybe I'm missing something, but this function is called only from here and only reads one line, so why can't you just pass the line you just read to it?

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+                   depth=depjma, depth_type="from location",
+                   region=region)
+    o_jma.resource_id = _get_resource_id(event_name,
+                                         'orgin', 'JMA')
```
origin?

yannikbehr replied:
?

QuLogic replied:
It currently says `'orgin'`; should be `'origin'`?

QuLogic commented on obspy/io/nied/fnetmt/core.py:
```diff
@@ -0,0 +1,258 @@
+                  depth=depmt, region=region,
+                  depth_type="from moment tensor inversion")
+    o_mt.resource_id = _get_resource_id(event_name,
+                                        'orgin', 'MT')
```
origin?

QuLogic commented on obspy/io/nied/fnetmt/util.py:
```diff
@@ -0,0 +1,38 @@
+    dx = (((370 * 24) * 60) * 60) * 1000
+    rng = 1
+    for _ in range(numenc):
+        rng *= numsym
```
`rng = numsym**numenc`?

QuLogic commented on obspy/io/nied/fnetmt/util.py:
```diff
@@ -0,0 +1,38 @@
+    numsym = len(sym)
+    x = (((((dt.julday - 1) * 24) + dt.hour) * 60 + dt.minute) *
+         60 + dt.second) * 1000 + dt.microsecond / 1000
+    dx = (((370 * 24) * 60) * 60) * 1000
```
370?

yannikbehr replied:
No idea why this is 370. I just took this from the SeisComP3 source code.

QuLogic commented on obspy/io/nied/fnetmt/util.py:
```diff
@@ -0,0 +1,38 @@
+        x = x * int(rng / dx)
+    enc = ''
+    for _ in range(numenc):
+        d = int(x / numsym)
```
Since `d` is unused, you could merge this line into 37.

QuLogic commented on obspy/io/nied/knet/__init__.py:
```diff
@@ -0,0 +1,116 @@
+problems if you want to write MiniSEED as it only allows 5 character station
+names. In this case you can opt to write the last 2 characters of the station
+name into the location field. This is possible because the location is encoded
+in the channel name (s.o.).
```
What is s.o.?

yannikbehr replied:
German abbreviation for 'see above' :-). Fixed.

QuLogic commented on obspy/io/nied/knet/__init__.py:
```diff
@@ -0,0 +1,116 @@
+1 Trace(s) in Stream:
+BO.AKT0.13.EW | 1996-08-10T18:12... - 1996-08-10T... | 100.0 Hz, 5900 samples
+
+Additional header entries from are written to an additional dictionary called
```
from what?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+from obspy.core.trace import Stats
+from datetime import datetime
+import re
+import numpy as np
```
Not really the right order (stdlib/third-party/first-party).

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    pass
+
+
+class KnetDataError(Exception):
```
Unused exception.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    :param buf: File to read.
+    :type buf: Open file or open file like object.
+    """
+    first_string = buf.read(11).decode()
```
Should it be 'ascii'? Or do they really mean not-binary?

yannikbehr replied:
I thought this defaults to utf-8?

QuLogic replied:
No, it defaults to system encoding, which may or may not be utf-8. And is the text really utf-8?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+        return line.split()
+
+
+def _read_knet_hdr(hdrlines, convert_stnm=False, **kwargs):
```
I don't think `kwargs` is used, and there's only one caller, so it shouldn't be necessary.

yannikbehr replied:
What would be a better way to pass through the `convert_stnm` keyword from `_read_knet_ascii`?

QuLogic replied:
But you don't call `_read_knet_ascii` from this function?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    # Event information
+    flds = _prep_hdr_line(hdrnames[_i], hdrlines[_i])
+    dt = flds[2] + ' ' + flds[3]
+    dt = UTCDateTime(datetime.strptime(dt, '%Y/%m/%d %H:%M:%S'))
```
Use `strptime` from `UTCDateTime`?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+        stnm = stnm[:-2]
+    if len(stnm) > 7:
+        raise KnetFormatError("Station name can't be more than 7 characters\
+        long!")
```
I prefer just starting the string indented on the next line, and I think it should all fit.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    flds = _prep_hdr_line(hdrnames[_i], hdrlines[_i])
+    dt = flds[2] + ' ' + flds[3]
+    dt = UTCDateTime(datetime.strptime(dt, '%Y/%m/%d %H:%M:%S'))
+    # All times are in Japanese standard time which is 9 hours ahead of GMT
```
A `UTCDateTime` should be UTC, not GMT, naturally.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    # A 15 s delay is added to the record time by the
+    # the K-NET and KiK-Net data logger
+    dt = UTCDateTime(datetime.strptime(dt, '%Y/%m/%d %H:%M:%S')) - 15.0
+    # All times are in Japanese standard time which is 9 hours ahead of GMT
```
A `UTCDateTime` should be UTC, not GMT, naturally.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    freq = int(m.group())
+    delta = 1.0 / freq
+    hdrdict['delta'] = delta
+    hdrdict['sampling_rate'] = freq
```
I don't think it's necessary to set both of these. They should be automatically related.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    flds = _prep_hdr_line(hdrnames[_i], hdrlines[_i])
+    dt = flds[2] + ' ' + flds[3]
+    dt = UTCDateTime(datetime.strptime(dt, '%Y/%m/%d %H:%M:%S'))
+    # All times are in Japanese standard time which is 9 hours ahead of GMT
```
A `UTCDateTime` should be UTC, not GMT, naturally.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+    if len(flds) > 1:
+        hdrdict['knet']['comment'] = ' '.join(flds[1:])
+
+    if len(hdrlines) != _i + 1:
```
Shouldn't this be checked at the beginning of the function? What if the number of header numbers is too short?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+            headerlines.append(line)
+            hdrdict = _read_knet_hdr(headerlines, **kwargs)
+            break
+        headerlines.append(line)
```
This could be moved above the test.

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+        data += [float(p) for p in parts]
+
+    hdrdict['npts'] = len(data)
+    elapsed = float(hdrdict['npts']) / float(hdrdict['sampling_rate'])
```
`float` shouldn't be necessary with `future`, no?

QuLogic commented on obspy/io/nied/knet/core.py:
```diff
@@ -0,0 +1,312 @@
+
+    hdrdict['npts'] = len(data)
+    elapsed = float(hdrdict['npts']) / float(hdrdict['sampling_rate'])
+    hdrdict['endtime'] = hdrdict['starttime'] + elapsed
```
This is set automatically.

QuLogic replied:
This is set automatically.

QuLogic commented on obspy/io/nied/tests/test_fnetmt_reading.py:
```diff
@@ -0,0 +1,80 @@
+
+class FNETMTCatalogReadingTestCase(unittest.TestCase):
+    """
+    Test everything related to the reading an F-net moment tensor catalog.
```
reading _of_

QuLogic commented on obspy/io/nied/tests/test_fnetmt_reading.py:
```diff
@@ -0,0 +1,80 @@
+
+    def test_read_fnetmt_catalog_from_bytes_io(self):
+        """
+        Tests that reading an F-net moment tensor file from an BytesIO objects
```
a BytesIO object

QuLogic commented on obspy/io/nied/tests/test_fnetmt_reading.py:
```diff
@@ -0,0 +1,80 @@
+
+import os
+import io
+import unittest
```
Newline after.

QuLogic replied:
(Fixed.)

QuLogic commented on obspy/io/nied/tests/test_knet_reading.py:
```diff
@@ -0,0 +1,108 @@
+import unittest
+from obspy import read
+from obspy.io.nied.knet.core import _is_knet_ascii
+import numpy as np
```
See above comment on import sorting.

QuLogic commented on obspy/io/nied/tests/test_knet_reading.py:
```diff
@@ -0,0 +1,108 @@
+
+    def test_read_knet_ascii_from_bytes_io(self):
+        """
+        Tests that reading an NIED moment tensor file from an BytesIO objects
```
a BytesIO object

QuLogic commented on obspy/io/nied/tests/test_knet_reading.py:
```diff
@@ -0,0 +1,108 @@
+
+    def test_station_name_hack(self):
+        """
+        Station names in K-NET and KiK-net are 6 characters long which is not
```
s/is/does/

ThomasLecocq commented:
@megies  @claudiodsf  indeed, the win32 is the latest format of the "datamark" format which I should have named "win" ...

yannikbehr commented:
@megies I thought that the directory nesting was what you had in mind by 'bundling'. 
@megies @ThomasLecocq @claudiodsf If the win format is used for both archived waveforms and a Japanese datalogger the name `nied` for the overarching module probable doesn't make much sense. Shall we revert the nested file structure?

yannikbehr commented:
@QuLogic thanks for going through the code.

megies commented:
> @megies I thought that the directory nesting was what you had in mind by 'bundling'.

I just thought we probably don't need the extra "stub" directories.. if I had to do it, I would have probably made it..
```
obspy/io/nied/fnetmt.py
obspy/io/nied/knet.py
```
instead of
```
obspy/io/nied/fnetmt/core.py
obspy/io/nied/knet/core.py
```
since there's not much besides the `core.py`s the directories are just "stubs". But that's just cosmetics really, can also stay as is.

---

One more thing:
The only thing that is not in one of the `core.py`s is `gen_sc3_id(t, ...)`. I think this logically belongs in `core`, probably I would put it into `core/utcdatetime.py`. I think, we should always make sure to move general helper routines not specific to the submodule into `core`, to increase visibility/reusability for later works. Since we have no `seiscomp` related place in obspy, that's were it fits best. Maybe it's even worth a thought to attach it to `UTCDateTime` as a method, only opposing argument is to avoid clutter of its namespace with highly special-purpose functionality (but I would still probably attach it as `UTCDateTime.get_seiscomp3_id(...)`).

---

> @megies @ThomasLecocq @claudiodsf If the win format is used for both archived waveforms and a Japanese datalogger the name nied for the overarching module probable doesn't make much sense. Shall we revert the nested file structure?

I think nobody is 100% happy with `obspy.io.nied`. I really don't have much insight in this naming maze (splotch of names of agencies/dataloggers/company names/you-name-it). If there's a better option please name it. :-)

ThomasLecocq commented:
@yannikbehr well... the Datamark dataloggers have been sold in many places, like in Indonesia... So I am not sure about nesting it under the /agency/ thingy...

QuLogic commented:
There is an issue with the history here. There are maybe 33 or so duplicate commits. It looks like you pulled from @krischer's repo and merged, but he did a rebase, so you needed to rebase any new changes on top of that.

megies commented:
Tried to fix the history several times now but failed, hard to tell for me what is going on with the history here anyway and it's 180 commits... Unless somebody is willing to put more time into fixing the history (or even able to do so), I'm tempted to squash everything into one single commit.

megies commented:
Spent roughly an hour trying  to clean this history up, this is what I came up with after two runs of rebasing: https://github.com/obspy/obspy/tree/japan_rebase
Lacking enthusiasm to verify now.

Unless anybody else is willing to put more time into history cleanup (I have no idea what was done to get it this way or where/how it went wrong) I only see a squash of the whole PR into a single commit.

yannikbehr commented:
@megies Thanks for working on this. I had it on my todo list the whole week but just didn't get around to look into it. I won't have time until Monday but will try to spend some time on this then. Is there a deadline when this should be finished?

megies commented:
> Is there a deadline when this should be finished?

We wanted to do the next major release 1-2 weeks before AGU, so I would like to be release-ready next week..

> I had it on my todo list the whole week but just didn't get around to look into it.

Do you recall what went wrong with the history and how? If I know how it got in this state I might be able to fix it..

yannikbehr commented:
I've studied the history with gitk for a while but I still don't know what went wrong. Somehow some of the commits (e.g. c6014f2c712bb8a360a21141980d1bb70bee1b0f to b98fbf3cd487f0d95d35155038271595fb4072ba ) were committed again when I rebased with the github repo. I guess it has to do with me doing something wrong when merging Lion's pull request. I tried to get rid of the multiple commits using `git rebase --onto` commands but without success. @QuLogic do you have an idea how the history could be fixed? I guess I could revert the history and then cherry pick the single commits but I think in this case I'd prefer squashing everything into 1 commit.

krischer commented:
I would be fine with squashing everything. Seems to be nasty to fix the history and (while it would be nice to have) its not worth the required effort IMHO.

QuLogic commented:
It's not terribly difficult to rebase; mostly just drop the first copy of the commits, and move the few before the first duplicate to be somewhere around the merge point. I don't think there are any reasons not to squash though. However, I think 8fc17a1f32b494f33b85059feff9a2d1204c7535 and part of 8bc507b6e6691828569cbef49f2ea64c49c9eea5 should be plucked separately. See [here](https://github.com/obspy/obspy/tree/japan_squash).

QuLogic commented:
@yannikbehr Do you want to update this PR, or should I just open a new one with the squashed commits?

yannikbehr commented:
@QuLogic @megies @krischer I've now cleaned the history from duplicate commits with `git rebase -i` and while doing that also removed the following commits from the history: 58a5ad5 438ebb9 b692f10 82a68c9 d60a7dc 0fb1921 333f230 6591f2f 1ca0a31 804b1f4 414ec8c 9051222 30cb3fb 8bc507b 8fc17a1

I committed the whole thing to [here](https://github.com/yannikbehr/obspy/tree/japan_clean)
Can you have a quick look at it to see whether this is OK?

megies commented:
I just took this branch here and `japan_clean`, merged current `upstream/master` into each and compared the dir trees with `diff`.
Looks mostly good, only a handful of changes that dropped out in the rebase it seems..?
I also took `japan_clean` and rebased it onto the most up-to-date `upstream/master` (only a handful of manual merge conflicts that needed solving), pushed the results to [megies/japan_clean_rebased](https://github.com/megies/obspy/commits/japan_clean_rebased). I think we can take that and add those changes below on top again, and it should be fine..

```diff
diff -r japan_master/misc/docs/source/conf.py japan_clean_master/misc/docs/source/conf.py
16d15
< READ_THE_DOCS = os.environ.get('READTHEDOCS', None) == 'True'
18,21c17
< # Force the use the agg backend which somehow on some systems is not selected
< # by the sphinx plotting routine.
< import matplotlib.pyplot as plt
< plt.switch_backend("agg")
---
> READ_THE_DOCS = os.environ.get('READTHEDOCS', None) == 'True'
Only in japan_master/misc/docs/source/packages: obspy.io.knet.rst
Only in japan_master/misc/docs/source/packages: obspy.io.nied.rst
diff -r japan_master/obspy/io/nied/fnetmt.py japan_clean_master/obspy/io/nied/fnetmt.py
99a100
>             line = line.decode()
151c152
<         line = buf.readline()
---
>         line = buf.readline().decode()
167c168
<             events.append(__read_single_fnetmt_entry(line))
---
>             events.append(__read_single_fnetmt_entry(line.decode()))
diff -r japan_master/obspy/io/nied/__init__.py japan_clean_master/obspy/io/nied/__init__.py
25c25
< 2011-03-11T05:46:18.120000Z | +38.103, +142.861 | 9.0 ML
---
> 2011-03-11T05:46:18.120000Z | +38.103, +142.861 | 8.7 Mw
30c30
< Event:  2011-03-11T05:46:18.120000Z | +38.103, +142.861 | 9.0 ML
---
> Event:  2011-03-11T05:46:18.120000Z | +38.103, +142.861 | 8.7 Mw
diff -r japan_master/obspy/io/nied/knet.py japan_clean_master/obspy/io/nied/knet.py
75c75
<     first_string = buf.read(11)
---
>     first_string = buf.read(11).decode()
276c276
<         line = buf.readline()
---
>         line = buf.readline().decode()
diff -r japan_master/obspy/io/nied/tests/test_knet_reading.py japan_clean_master/obspy/io/nied/tests/test_knet_reading.py
52,53c52,53
<         Tests that reading an NIED moment tensor file from a BytesIO object
<         works.
---
>         Tests that reading of K-NET and KiK-net ASCII format files from a
>         BytesIO object works.
```

yannikbehr commented:
@megies Only the missing files `obspy.io.knet.rst` and `obspy.io.nied.rst` need to be fixed, the rest of the diffs is intentional. I added the `.decode()` statements back in as I saw that Lion had added them for Python3 compatibility (f684c5e 51efd92). Some diffs are due to new commits (0905842 0384cc8) and then there is one commit that Elliott wanted to commit separately (8fc17a1).  So it should suffice to add the two `.rst` files and 8fc17a1.

QuLogic commented:
Looking at the build results from the squashed version, it makes sense to be adding those `decode` back in for Python 3. You do not need to leave 8fc17a1 out of this PR specifically, only if you squash it all then it should be a separate commit.

You will still need to rebase again onto the latest master because there are merge conflicts.

megies commented:
@yannikbehr, OK, since you got an overview of what still needs doing I would propose to force-push https://github.com/megies/obspy/commits/japan_clean_rebased to this branch here. Then you can do the final touches and we should be good I guess.. ?



> You will still need to rebase again onto the latest master because there are merge conflicts.

Did that already with `megies/japan_clean_rebased`, see above.

yannikbehr commented:
@megies OK, done.

megies commented:
@yannikbehr, this file shouldnt be in there, right? https://github.com/obspy/obspy/pull/1125/files#diff-2c5db54bb866f042d1c6a29a61531091

yannikbehr commented:
@megies true. removed.

megies commented:
Two trivial PEP8 issues left.. https://travis-ci.org/obspy/obspy/jobs/90770054#L808

QuLogic commented on obspy/io/nied/fnetmt.py:
```diff
@@ -0,0 +1,254 @@
+                   events=events, description=headerlines[:-1])
+
+
+def __read_single_fnetmt_entry(line, **kwargs):
```
I still don't know what the point of the `kwargs` is here; only one place calls this function and it never passes in any keyword arguments.

QuLogic commented on obspy/io/nied/knet.py:
```diff
@@ -0,0 +1,300 @@
+    :param hdrlines: List of the header lines of a a K-NET/KiK-net ASCII file
+    :type hdrlines: list
+    :param convert_stnm: For station names with 6 letters write the last two
+    letters of the station code to the 'location' field
```
I think this needs to be indented since it's wrapper.

megies replied:
:+1:

QuLogic commented on obspy/io/nied/knet.py:
```diff
@@ -0,0 +1,300 @@
+        return line.split()
+
+
+def _read_knet_hdr(hdrlines, convert_stnm=False, **kwargs):
```
Another `kwargs` which I don't see being used.

QuLogic replied:
Oh, never mind, I see it's so `convert_stnm` can be expanded while ignoring any other keywords.

QuLogic commented on misc/docs/source/packages/obspy.io.nied.fnetmt.rst:
```diff
@@ -0,0 +1,15 @@
+       :toctree: autogen
+       :nosignatures:
+
+       core
```
There is no `core` within `obspy.io.nied.fnetmt`. In fact, I'm not sure you even need this file any more.

QuLogic commented on misc/docs/source/packages/obspy.io.nied.knet.rst:
```diff
@@ -0,0 +1,14 @@
+       :toctree: autogen
+       :nosignatures:
+
+       core
```
There is no `core` within `obspy.io.nied.knet`. In fact, I'm not sure you even need this file any more.

QuLogic commented on misc/docs/source/packages/obspy.io.nied.rst:
```diff
@@ -0,0 +1,14 @@
+       :toctree: autogen
+       :nosignatures:
+
+       core
```
There is no `core` within `obspy.io.nied`. In fact, I think you need to list `fnetmt` and `knet` here and drop those other two files.

megies replied:
Nice catch, I agree.

QuLogic commented on obspy/io/nied/__init__.py:
```diff
@@ -0,0 +1,152 @@
+"""
+obspy.io.nied.fnetmt - F-net's moment tensors TEXT format support for ObsPy
+===================================================================
```
The underline is too short.

QuLogic commented on obspy/io/nied/__init__.py:
```diff
@@ -0,0 +1,152 @@
+               magnitudes: 2 Elements
+
+obspy.io.nied.knet - K-NET and KiK-net ASCII format support for ObsPy
+================================================================
```
The underline is too short.

QuLogic commented:
OK, I tried to also check for any documentation issues as well, but `master` is throwing 15K warnings, so it's hard to say that nothing is wrong.

megies commented:
Nobody came up with a better fitting name than `io.nied` so I guess this is good to merge now?

krischer commented:
:ship:

yannikbehr commented:
:japanese_ogre: :japan:

megies commented:
:bouquet:

claudiodsf commented:
:clap:
